source/common/stats/symbol_table_impl.h

```diff
@@ -0,0 +1,138 @@
+#pragma once
+
+#include <algorithm>
+#include <memory>
+#include <stack>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "envoy/common/exception.h"
+#include "envoy/stats/symbol_table.h"
+
+#include "common/common/assert.h"
+#include "common/common/utility.h"
+
+#include "absl/strings/str_join.h"
+#include "absl/strings/str_split.h"
+
+namespace Envoy {
+namespace Stats {
+
+using Symbol = uint32_t;
+using SymbolVec = std::vector<Symbol>;
+
+/**
+ * Underlying SymbolTableImpl implementation which manages per-symbol reference counting.
+ *
+ * The underlying Symbol / SymbolVec data structures are private to the impl. One side
+ * effect of the non-monotonically-increasing symbol counter is that if a string is encoded, the
+ * resulting stat is destroyed, and then that same string is re-encoded, it may or may not encode to
+ * the same underlying symbol.
+ */
+class SymbolTableImpl : public SymbolTable {
+public:
+  StatNamePtr encode(const absl::string_view name) override;
+
+  // For testing purposes only.
+  size_t size() const override {
+    ASSERT(encode_map_.size() == decode_map_.size());
+    return encode_map_.size();
+  }
+
+private:
+  friend class StatNameImpl;
+  friend class StatNameTest;
+
+  struct SharedSymbol {
+    Symbol symbol_;
+    uint32_t ref_count_;
+  };
+
+  /**
+   * Decodes a vector of symbols back into its period-delimited stat name.
+   * If decoding fails on any part of the symbol_vec, we release_assert and crash hard, since this
+   * should never happen, and we don't want to continue running with a corrupt stats set.
+   *
+   * @param symbol_vec the vector of symbols to decode.
+   * @return std::string the retrieved stat name.
+   */
+  std::string decode(const SymbolVec& symbol_vec) const;
+
+  /**
+   * Since SymbolTableImpl does manual reference counting, a client of SymbolTable (such as
+   * StatName) must manually call free(symbol_vec) when it is freeing the stat it represents. This
+   * way, the symbol table will grow and shrink dynamically, instead of being write-only.
+   *
+   * @param symbol_vec the vector of symbols to be freed.
+   */
+  void free(const SymbolVec& symbol_vec);
+
+  /**
+   * Convenience function for encode(), symbolizing one string segment at a time.
+   *
+   * @param sv the individual string to be encoded as a symbol.
+   * @return Symbol the encoded string.
+   */
+  Symbol toSymbol(absl::string_view sv);
+
+  /**
+   * Convenience function for decode(), decoding one symbol at a time.
+   *
+   * @param symbol the individual symbol to be decoded.
+   * @return absl::string_view the decoded string.
+   */
+  absl::string_view fromSymbol(const Symbol symbol) const;
+
+  // Stages a new symbol for use. To be called after a successful insertion.
+  void newSymbol() {
+    if (pool_.empty()) {
+      ASSERT(monotonic_counter_ < std::numeric_limits<Symbol>::max());
```

Comment:
Won't this be true by type properties?

Reply:
I think `counter <= max` will always be true by type properties, but the purpose of this line was to catch it when it hits `counter == max` and assert-crash Envoy, to avoid overflow or other weirdness. What's the proper way to handle that case and crash (or recover, if you can think of a way) ?

Reply:
You could also check for it wrapping to zero after you increment it.

Reply:
That sounds more explicit, I'll change it.